File: src/main/java/de/fhopf/akka/VisitedPageStore.java
package de.fhopf.akka;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * Stores the information on which pages have already been retrieved.
 * @author Florian Hopf, http://www.florian-hopf.de
 */
public class VisitedPageStore {

    private Set<String> pagesToVisit = new HashSet<String>();
    private Set<String> allPages = new HashSet<String>();
    private Set<String> inProgress = new HashSet<String>();
    
    public void add(String page) {
        if (!allPages.contains(page)) {
            pagesToVisit.add(page);
            allPages.add(page);
        }
    }
    
    public void addAll(Collection<String> pages) {
        for (String page: pages) {
            add(page);
        }
    }
    
    public void finished(String page) {
        inProgress.remove(page);
    }
    
    public String getNext() {
        if (pagesToVisit.isEmpty()) {
            return null;
        } else {
            String next = pagesToVisit.iterator().next();
            pagesToVisit.remove(next);
            inProgress.add(next);
            return next;
        }
    }
    
    public Collection<String> getNextBatch() {
        Set<String> pages = new HashSet<String>();
        pages.addAll(pagesToVisit);
        pagesToVisit.clear();
        inProgress.addAll(pages);
        return pages;
    }
    
    public boolean isFinished() {
        return pagesToVisit.isEmpty() && inProgress.isEmpty();
    }
    
    @Override
    public String toString() {
        return String.format("inProgress: %1$3s, allPages: %2$3s", inProgress.size(), allPages.size());
    }

}

File: src/main/scala/de/fhopf/akka/VisitedPageStore.scala
package de.fhopf.akka

import java.util.Collection
import java.util.HashSet
import java.util.Set
import collection.JavaConversions._

/**
 * @author Florian Hopf, http://www.florian-hopf.de
 * @author Alfredo Serafini, http://www.seralf.it
 */
class VisitedPageStore {

  private var pagesToVisit: Set[String] = new HashSet[String]()
  private var allPages: Set[String] = new HashSet[String]()
  private var inProgress: Set[String] = new HashSet[String]()

  def add(page: String) = {
    if (!allPages.contains(page)) {
      pagesToVisit.add(page)
      allPages.add(page)
    }
  }

  def addAll(pages: Collection[String]) = {
    for (page <- pages) {
      add(page)
    }
  }

  def finished(page: String) = {
    inProgress.remove(page)
  }

  def getNext() = {
    if (pagesToVisit.isEmpty()) {
      null
    } else {
      val next: String = pagesToVisit.iterator().next()
      pagesToVisit.remove(next)
      inProgress.add(next)
      next
    }
  }

  def getNextBatch(): Collection[String] = {
    val pages: Set[String] = new HashSet[String]()
    pages.addAll(pagesToVisit)
    pagesToVisit.clear()
    inProgress.addAll(pages)
    pages
  }

  def isFinished(): Boolean = pagesToVisit.isEmpty() && inProgress.isEmpty()

}
